src/core/gltf2importer/meshparser.cpp

```diff
@@ -123,6 +125,201 @@ Qt3DRender::QAttribute *decodeAttribute(const draco::PointCloud *pointCould,
+        vertexIndices[0] = typedIndices[0];
+        vertexIndices[1] = typedIndices[1];
+        vertexIndices[2] = typedIndices[2];
+        return vertexIndices;
+    }
+    }
+}
+
+// TODO This is only needed when the material has normal mapping
+// TODO Compute tangents when the primitive is not indexed
+Qt3DRender::QAttribute *createTangentAttribute(Qt3DRender::QGeometry *geometry)
+{
+    struct MikkTSpaceUserData
+    {
+        const Qt3DRender::QAttribute *indexAttribute = nullptr;
+        const Qt3DRender::QAttribute *positionAttribute = nullptr;
+        const Qt3DRender::QAttribute *normalAttribute = nullptr;
+        const Qt3DRender::QAttribute *uvAttribute = nullptr;
+        Qt3DRender::QAttribute *tangentAttribute = nullptr;
+    };
+
+    MikkTSpaceUserData userData;
+    const auto &attributes = geometry->attributes();
+    std::for_each(std::begin(attributes), std::end(attributes), [&userData] (const Qt3DRender::QAttribute *attr) {
+        if (attr->attributeType() == Qt3DRender::QAttribute::IndexAttribute)
+            userData.indexAttribute = attr;
+        if (attr->name() == Qt3DRender::QAttribute::defaultPositionAttributeName())
+            userData.positionAttribute = attr;
+        if (attr->name() == Qt3DRender::QAttribute::defaultNormalAttributeName())
+            userData.normalAttribute = attr;
+        if (attr->name() == Qt3DRender::QAttribute::defaultTextureCoordinateAttributeName())
+            userData.uvAttribute = attr;
+    });
+
+    if (!userData.indexAttribute ||
+        !userData.positionAttribute ||
+        !userData.normalAttribute ||
+        !userData.uvAttribute)
+        return nullptr;
+
+    // Create the tangent attribute
+    QByteArray tangentBufferData;
+    tangentBufferData.resize(userData.positionAttribute->count()*sizeof(QVector4D));
+    auto *tangentBuffer = new Qt3DRender::QBuffer;
+    tangentBuffer->setData(tangentBufferData);
+    auto *tangentAttribute = new Qt3DRender::QAttribute(tangentBuffer,
+                                                        Qt3DRender::QAttribute::Float,
+                                                        4,
+                                                        userData.positionAttribute->count());
+    userData.tangentAttribute = tangentAttribute;
+
+    //Create callbacks for MikkTSpace
+    SMikkTSpaceContext mikkContext;
+    mikkContext.m_pUserData = &userData;
+
+    SMikkTSpaceInterface interface;
+    interface.m_getNumFaces = [](const SMikkTSpaceContext *pContext) -> int {
+        const auto *userData = reinterpret_cast<MikkTSpaceUserData*>(pContext->m_pUserData);
+        return userData->indexAttribute->count() / 3;
+    };
+
+    interface.m_getNumVerticesOfFace = [](const SMikkTSpaceContext *pContext, const int iFace) ->int {
+        return 3;
+    };
+
+    interface.m_getPosition = [] (const SMikkTSpaceContext *pContext,
+                    float fvPosOut[],
```

Comment:
Indentation looks wrong here

Reply:
could be the array declaration in the lambda parameters that threw the auto-indenter out ?